File: src/cubyz/client/BlockMeshes.java
package cubyz.client;

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

import javax.imageio.ImageIO;

import cubyz.utils.Logger;
import cubyz.utils.datastructures.IntFastList;
import cubyz.api.DataOrientedRegistry;
import cubyz.api.Resource;
import cubyz.rendering.Material;
import cubyz.rendering.Mesh;
import cubyz.rendering.ModelLoader;
import cubyz.rendering.SSBO;
import cubyz.rendering.Texture;
import cubyz.rendering.TextureArray;
import cubyz.utils.json.JsonElement;
import cubyz.utils.json.JsonObject;
import cubyz.utils.json.JsonString;
import cubyz.world.Neighbors;
import cubyz.world.blocks.Blocks;

public class BlockMeshes implements DataOrientedRegistry {

	private static int size = 1;
	private static final Mesh[] meshes = new Mesh[Blocks.MAX_BLOCK_COUNT];
	private static final String[] models = new String[Blocks.MAX_BLOCK_COUNT];
	private static final int[][] textureIndices = new int[Blocks.MAX_BLOCK_COUNT][6];
	/** Stores the number of textures after each block was added. Used to clean additional textures when the world is switched.*/
	private static final int[] maxTextureCount = new int[Blocks.MAX_BLOCK_COUNT];
	/** Number of loaded meshes. Used to determine if an update is needed */
	private static int loadedMeshes = 1;

	private static ArrayList<BufferedImage> blockTextures = new ArrayList<BufferedImage>();
	private static IntFastList animationFrames = new IntFastList(8192);
	private static IntFastList animationTimes = new IntFastList(8192);
	private static ArrayList<String> textureIDs = new ArrayList<String>();

	private static final String[] sideNames = new String[6];

	private static SSBO animationTimesSSBO;
	private static SSBO animationFramesSSBO;

	static {
		sideNames[Neighbors.DIR_DOWN] = "bottom";
		sideNames[Neighbors.DIR_UP] = "top";
		sideNames[Neighbors.DIR_POS_X] = "right";
		sideNames[Neighbors.DIR_NEG_X] = "left";
		sideNames[Neighbors.DIR_POS_Z] = "front";
		sideNames[Neighbors.DIR_NEG_Z] = "back";

		readTexture(new JsonString("cubyz:undefined"), "assets/");

		animationTimesSSBO = new SSBO(0);
		animationFramesSSBO = new SSBO(1);
	}

	public static Mesh mesh(int block) {
		return meshes[block & Blocks.TYPE_MASK];
	}
	public static int[] textureIndices(int block) {
		return textureIndices[block & Blocks.TYPE_MASK];
	}

	public static int readTexture(JsonElement textureInfo, String assetFolder) {
		int result = -1;
		if (textureInfo instanceof JsonString) {
			String resource = textureInfo.getStringValue(null);
			if (resource != null) {
				Resource texture = new Resource(resource);
				String path = assetFolder + texture.getMod() + "/blocks/textures/" + texture.getID() + ".png";
				// Test if it's already in the list:
				for(int j = 0; j < textureIDs.size(); j++) {
					if (textureIDs.get(j).equals(path)) {
						result = j;
						return result;
					}
				}
				// Otherwise read it into the list:
				result = blockTextures.size();
				try {
					blockTextures.add(ImageIO.read(new File(path)));
					textureIDs.add(path);
					animationFrames.add(1);
					animationTimes.add(1);
				} catch(IOException e) {
					result = -1;
					Logger.warning("Could not read image "+texture+" from Block "+Blocks.id(size));
					Logger.warning(e);
				}
			}
		} else if (textureInfo instanceof JsonObject) {
			int animationTime = textureInfo.getInt("time", 500);
			String[] textures = textureInfo.getArrayNoNull("textures").getStrings();
			// Add the new textures into the list. Since this is an animation all textures that weren't found need to be replaced with undefined.
			result = blockTextures.size();
			for(int i = 0; i < textures.length; i++) {
				if (i == 0) {
					animationFrames.add(textures.length);
					animationTimes.add(animationTime);
				} else {
					animationFrames.add(1);
					animationTimes.add(1);
				}
				Resource texture = new Resource(textures[i]);
				try {
					String path = assetFolder + texture.getMod() + "/blocks/textures/" + texture.getID() + ".png";
					blockTextures.add(ImageIO.read(new File(path)));
					if (i == 0) {
						textureIDs.add(path);
					} else {
						textureIDs.add(path+":animation");
					}
				} catch(IOException e) {
					Logger.warning("Could not read image "+texture+" from Block "+Blocks.id(size));
					Logger.warning(e);
					blockTextures.add(blockTextures.get(0));
					textureIDs.add(textureIDs.get(0));
				}
			}
		}
		return result;
	}

	public static void getTextureIndices(JsonObject json, String assetFolder, int[] textureIndices) {
		for(int i = 0; i < 6; i++) {
			JsonElement textureInfo = json.get("texture_"+sideNames[i]);
			textureIndices[i] = readTexture(textureInfo, assetFolder);
		}

		int remainingIndex = readTexture(json.get("texture"), assetFolder);
		if (remainingIndex == -1) remainingIndex = 0;
		for(int i = 0; i < 6; i++) {
			if (textureIndices[i] == -1)
				textureIndices[i] = remainingIndex;
		}
	}

	@Override
	public Resource getRegistryID() {
		return new Resource("cubyz:block_meshes");
	}

	@Override
	public int register(String assetFolder, Resource id, JsonObject json) {
		models[size] = json.getString("model", "cubyz:block.obj");

		// The actual model is loaded later, in the rendering thread.
		// But textures can be loaded here:

		getTextureIndices(json, assetFolder, textureIndices[size]);
		
		maxTextureCount[size] = textureIDs.size();

		return size++;
	}

	@Override
	public void reset(int len) {
		for(int i = len; i < size; i++) {
			meshes[i] = null;
			models[i] = null;
		}
		size = len;
		loadedMeshes = len;
		for(int i = textureIDs.size() - 1; i >= maxTextureCount[size-1]; i--) {
			textureIDs.remove(i);
			blockTextures.remove(i);
		}
	}

	public static void reloadTextures() {
		for(int i = 0; i < blockTextures.size(); i++) {
			try {
				blockTextures.set(i, ImageIO.read(new File(textureIDs.get(i).replace(":animation", ""))));
			} catch(IOException e) {
				Logger.warning("Could not read image from path "+textureIDs.get(i));
				Logger.warning(e);
				blockTextures.set(i, blockTextures.get(0));
			}
		}
		generateTextureArray();
	}

	public static void loadMeshes() {
		// Goes through all meshes that were newly added:
		for(; loadedMeshes < size; loadedMeshes++) {
			if (meshes[loadedMeshes] == null) {
				meshes[loadedMeshes] = Meshes.cachedDefaultModels.get(models[loadedMeshes]);
				if (meshes[loadedMeshes] == null) {
					Resource rs = new Resource(models[loadedMeshes]);
					meshes[loadedMeshes] = new Mesh(ModelLoader.loadModel(rs, "assets/" + rs.getMod() + "/models/3d/" + rs.getID()));
					meshes[loadedMeshes].setMaterial(new Material((Texture)null, 0.6f));
					Meshes.cachedDefaultModels.put(models[loadedMeshes], meshes[loadedMeshes]);
				}
			}
		}
	}

	public static void generateTextureArray() {
		TextureArray textures = Meshes.blockTextureArray;
		textures.clear();
		for(int i = 0; i < blockTextures.size(); i++) {
			BufferedImage img = blockTextures.get(i);
			textures.addTexture(img);
		}
		textures.generate();


		// Also generate additional buffers:
		animationTimes.trimToSize();
		animationTimesSSBO.bufferData(animationTimes.array);
		
		animationFrames.trimToSize();
		animationFramesSSBO.bufferData(animationFrames.array);
	}
	
}

File: src/blocks/block_meshes.zig
const std = @import("std");

const main = @import("main");
const block_props = @import("block_props.zig");
const Block = block_props.Block;
const BlockProps = block_props.BlockProps;
const maxBlockCount = block_props.maxBlockCount;
const graphics = @import("../graphics.zig");
const SSBO = graphics.SSBO;
const Image = graphics.Image;
const Color = graphics.Color;
const TextureArray = graphics.TextureArray;
const models = @import("../models.zig");
const ModelIndex = models.ModelIndex;
const chunk = @import("../chunk.zig");
const Neighbor = chunk.Neighbor;
const ZonElement = @import("../zon.zig").ZonElement;


pub const meshes = struct {
	const AnimationData = extern struct {
		startFrame: u32,
		frames: u32,
		time: u32,
	};

	const FogData = extern struct {
		fogDensity: f32,
		fogColor: u32,
	};
	var size: u32 = 0;
	var _modelIndex: [maxBlockCount]ModelIndex = undefined;
	var textureIndices: [maxBlockCount][16]u16 = undefined;
	/// Stores the number of textures after each block was added. Used to clean additional textures when the world is switched.
	var maxTextureCount: [maxBlockCount]u32 = undefined;
	/// Number of loaded meshes. Used to determine if an update is needed.
	var loadedMeshes: u32 = 0;

	var textureIDs: main.List([]const u8) = undefined;
	var animation: main.List(AnimationData) = undefined;
	var blockTextures: main.List(Image) = undefined;
	var emissionTextures: main.List(Image) = undefined;
	var reflectivityTextures: main.List(Image) = undefined;
	var absorptionTextures: main.List(Image) = undefined;
	var textureFogData: main.List(FogData) = undefined;
	pub var textureOcclusionData: main.List(bool) = undefined;

	var arenaAllocatorForWorld: main.heap.NeverFailingArenaAllocator = undefined;
	var arenaForWorld: main.heap.NeverFailingAllocator = undefined;

	pub var blockBreakingTextures: main.List(u16) = undefined;

	const sideNames = blk: {
		var names: [6][]const u8 = undefined;
		names[Neighbor.dirDown.toInt()] = "texture_bottom";
		names[Neighbor.dirUp.toInt()] = "texture_top";
		names[Neighbor.dirPosX.toInt()] = "texture_right";
		names[Neighbor.dirNegX.toInt()] = "texture_left";
		names[Neighbor.dirPosY.toInt()] = "texture_front";
		names[Neighbor.dirNegY.toInt()] = "texture_back";
		break :blk names;
	};

	var animationSSBO: ?SSBO = null;
	var animatedTextureSSBO: ?SSBO = null;
	var fogSSBO: ?SSBO = null;

	var animationComputePipeline: graphics.ComputePipeline = undefined;
	var animationUniforms: struct {
		time: c_int,
		size: c_int,
	} = undefined;

	pub var blockTextureArray: TextureArray = undefined;
	pub var emissionTextureArray: TextureArray = undefined;
	pub var reflectivityAndAbsorptionTextureArray: TextureArray = undefined;
	pub var ditherTexture: graphics.Texture = undefined;

	const black: Color = Color{.r = 0, .g = 0, .b = 0, .a = 255};
	const magenta: Color = Color{.r = 255, .g = 0, .b = 255, .a = 255};
	var undefinedTexture = [_]Color{magenta, black, black, magenta};
	const undefinedImage = Image{.width = 2, .height = 2, .imageData = undefinedTexture[0..]};
	var emptyTexture = [_]Color{black};
	const emptyImage = Image{.width = 1, .height = 1, .imageData = emptyTexture[0..]};

	pub fn init() void {
		animationComputePipeline = graphics.ComputePipeline.init("assets/cubyz/shaders/animation_pre_processing.comp", "", &animationUniforms);
		blockTextureArray = .init();
		emissionTextureArray = .init();
		reflectivityAndAbsorptionTextureArray = .init();
		ditherTexture = .initFromMipmapFiles("assets/cubyz/blocks/textures/dither/", 64, 0.5);
		textureIDs = .init(main.globalAllocator);
		animation = .init(main.globalAllocator);
		blockTextures = .init(main.globalAllocator);
		emissionTextures = .init(main.globalAllocator);
		reflectivityTextures = .init(main.globalAllocator);
		absorptionTextures = .init(main.globalAllocator);
		textureFogData = .init(main.globalAllocator);
		textureOcclusionData = .init(main.globalAllocator);
		arenaAllocatorForWorld = .init(main.globalAllocator);
		arenaForWorld = arenaAllocatorForWorld.allocator();
		blockBreakingTextures = .init(main.globalAllocator);
	}

	pub fn deinit() void {
		if(animationSSBO) |ssbo| {
			ssbo.deinit();
		}
		if(animatedTextureSSBO) |ssbo| {
			ssbo.deinit();
		}
		if(fogSSBO) |ssbo| {
			ssbo.deinit();
		}
		animationComputePipeline.deinit();
		blockTextureArray.deinit();
		emissionTextureArray.deinit();
		reflectivityAndAbsorptionTextureArray.deinit();
		ditherTexture.deinit();
		textureIDs.deinit();
		animation.deinit();
		blockTextures.deinit();
		emissionTextures.deinit();
		reflectivityTextures.deinit();
		absorptionTextures.deinit();
		textureFogData.deinit();
		textureOcclusionData.deinit();
		arenaAllocatorForWorld.deinit();
		blockBreakingTextures.deinit();
	}

	pub fn reset() void {
		meshes.size = 0;
		loadedMeshes = 0;
		textureIDs.clearRetainingCapacity();
		animation.clearRetainingCapacity();
		blockTextures.clearRetainingCapacity();
		emissionTextures.clearRetainingCapacity();
		reflectivityTextures.clearRetainingCapacity();
		absorptionTextures.clearRetainingCapacity();
		textureFogData.clearRetainingCapacity();
		textureOcclusionData.clearRetainingCapacity();
		blockBreakingTextures.clearRetainingCapacity();
		_ = arenaAllocatorForWorld.reset(.free_all);
	}

	pub inline fn model(block: Block) ModelIndex {
		return block.mode().model(block);
	}

	pub inline fn modelIndexStart(block: Block) ModelIndex {
		return _modelIndex[block.typ];
	}

	pub inline fn fogDensity(block: Block) f32 {
		return textureFogData.items[animation.items[textureIndices[block.typ][0]].startFrame].fogDensity;
	}

	pub inline fn fogColor(block: Block) u32 {
		return textureFogData.items[animation.items[textureIndices[block.typ][0]].startFrame].fogColor;
	}

	pub inline fn hasFog(block: Block) bool {
		return fogDensity(block) != 0.0;
	}

	pub inline fn textureIndex(block: Block, orientation: usize) u16 {
		if(orientation < 16) {
			return textureIndices[block.typ][orientation];
		} else {
			return textureIndices[block.data][orientation - 16];
		}
	}

	fn extendedPath(_allocator: main.heap.NeverFailingAllocator, path: []const u8, ending: []const u8) []const u8 {
		return std.fmt.allocPrint(_allocator.allocator, "{s}{s}", .{path, ending}) catch unreachable;
	}

	fn readTextureFile(_path: []const u8, ending: []const u8, default: Image) Image {
		const path = extendedPath(main.stackAllocator, _path, ending);
		defer main.stackAllocator.free(path);
		return Image.readFromFile(arenaForWorld, path) catch default;
	}

	fn extractAnimationSlice(image: Image, frame: usize, frames: usize) Image {
		if(image.height < frames) return image;
		var startHeight = image.height/frames*frame;
		if(image.height%frames > frame) startHeight += frame else startHeight += image.height%frames;
		var endHeight = image.height/frames*(frame + 1);
		if(image.height%frames > frame + 1) endHeight += frame + 1 else endHeight += image.height%frames;
		var result = image;
		result.height = @intCast(endHeight - startHeight);
		result.imageData = result.imageData[startHeight*image.width .. endHeight*image.width];
		return result;
	}

	fn readTextureData(_path: []const u8) void {
		const path = _path[0 .. _path.len - ".png".len];
		const textureInfoPath = extendedPath(main.stackAllocator, path, ".zig.zon");
		defer main.stackAllocator.free(textureInfoPath);
		const textureInfoZon = main.files.cwd().readToZon(main.stackAllocator, textureInfoPath) catch .null;
		defer textureInfoZon.deinit(main.stackAllocator);
		const animationFrames = textureInfoZon.get(u32, "frames", 1);
		const animationTime = textureInfoZon.get(u32, "time", 1);
		animation.append(.{.startFrame = @intCast(blockTextures.items.len), .frames = animationFrames, .time = animationTime});
		const base = readTextureFile(path, ".png", Image.defaultImage);
		const emission = readTextureFile(path, "_emission.png", Image.emptyImage);
		const reflectivity = readTextureFile(path, "_reflectivity.png", Image.emptyImage);
		const absorption = readTextureFile(path, "_absorption.png", Image.whiteEmptyImage);
		for(0..animationFrames) |i| {
			blockTextures.append(extractAnimationSlice(base, i, animationFrames));
			emissionTextures.append(extractAnimationSlice(emission, i, animationFrames));
			reflectivityTextures.append(extractAnimationSlice(reflectivity, i, animationFrames));
			absorptionTextures.append(extractAnimationSlice(absorption, i, animationFrames));
			textureFogData.append(.{
				.fogDensity = textureInfoZon.get(f32, "fogDensity", 0.0),
				.fogColor = textureInfoZon.get(u32, "fogColor", 0xffffff),
			});
		}
		textureOcclusionData.append(textureInfoZon.get(bool, "hasOcclusion", true));
	}

	pub fn readTexture(_textureId: ?[]const u8, assetFolder: []const u8) !u16 {
		const textureId = _textureId orelse return error.NotFound;
		var result: u16 = undefined;
		var splitter = std.mem.splitScalar(u8, textureId, ':');
		const mod = splitter.first();
		const id = splitter.rest();
		var path = try std.fmt.allocPrint(main.stackAllocator.allocator, "{s}/{s}/blocks/textures/{s}.png", .{assetFolder, mod, id});
		defer main.stackAllocator.free(path);
		// Test if it's already in the list:
		for(textureIDs.items, 0..) |other, j| {
			if(std.mem.eql(u8, other, path)) {
				result = @intCast(j);
				return result;
			}
		}
		const file = main.files.cwd().openFile(path) catch |err| blk: {
			if(err != error.FileNotFound) {
				std.log.err("Could not open file {s}: {s}", .{path, @errorName(err)});
			}
			main.stackAllocator.free(path);
			path = try std.fmt.allocPrint(main.stackAllocator.allocator, "assets/{s}/blocks/textures/{s}.png", .{mod, id}); // Default to global assets.
			break :blk main.files.cwd().openFile(path) catch |err2| {
				std.log.err("File not found. Searched in \"{s}\" and also in the assetFolder \"{s}\"", .{path, assetFolder});
				return err2;
			};
		};
		file.close(); // It was only openend to check if it exists.
		// Otherwise read it into the list:
		result = @intCast(textureIDs.items.len);

		textureIDs.append(arenaForWorld.dupe(u8, path));
		readTextureData(path);
		return result;
	}

	pub fn getTextureIndices(zon: ZonElement, assetFolder: []const u8, textureIndicesRef: *[16]u16) void {
		const defaultIndex = readTexture(zon.get(?[]const u8, "texture", null), assetFolder) catch 0;
		inline for(textureIndicesRef, 0..) |*ref, i| {
			var textureId = zon.get(?[]const u8, std.fmt.comptimePrint("texture{}", .{i}), null);
			if(i < sideNames.len) {
				textureId = zon.get(?[]const u8, sideNames[i], textureId);
			}
			ref.* = readTexture(textureId, assetFolder) catch defaultIndex;
		}
	}

	pub fn register(assetFolder: []const u8, _: []const u8, zon: ZonElement) void {
		_modelIndex[meshes.size] = BlockProps.mode[meshes.size].createBlockModel(.{.typ = @intCast(meshes.size), .data = 0}, &BlockProps.modeData[meshes.size], zon.getChild("model"));

		// The actual model is loaded later, in the rendering thread.
		// But textures can be loaded here:

		getTextureIndices(zon, assetFolder, &textureIndices[meshes.size]);

		maxTextureCount[meshes.size] = @intCast(textureIDs.items.len);

		meshes.size += 1;
	}

	pub fn registerBlockBreakingAnimation(assetFolder: []const u8) void {
		var i: usize = 0;
		while(true) : (i += 1) {
			const path1 = std.fmt.allocPrint(main.stackAllocator.allocator, "assets/cubyz/blocks/textures/breaking/{}.png", .{i}) catch unreachable;
			defer main.stackAllocator.free(path1);
			const path2 = std.fmt.allocPrint(main.stackAllocator.allocator, "{s}/cubyz/blocks/textures/breaking/{}.png", .{assetFolder, i}) catch unreachable;
			defer main.stackAllocator.free(path2);
			if(!main.files.cwd().hasFile(path1) and !main.files.cwd().hasFile(path2)) break;

			const id = std.fmt.allocPrint(main.stackAllocator.allocator, "cubyz:breaking/{}", .{i}) catch unreachable;
			defer main.stackAllocator.free(id);
			blockBreakingTextures.append(readTexture(id, assetFolder) catch break);
		}
	}

	pub fn preProcessAnimationData(time: u32) void {
		animationComputePipeline.bind();
		graphics.c.glUniform1ui(animationUniforms.time, time);
		graphics.c.glUniform1ui(animationUniforms.size, @intCast(animation.items.len));
		graphics.c.glDispatchCompute(@intCast(@divFloor(animation.items.len + 63, 64)), 1, 1); // TODO: Replace with @divCeil once available
		graphics.c.glMemoryBarrier(graphics.c.GL_SHADER_STORAGE_BARRIER_BIT);
	}

	pub fn reloadTextures(_: usize) void {
		blockTextures.clearRetainingCapacity();
		emissionTextures.clearRetainingCapacity();
		reflectivityTextures.clearRetainingCapacity();
		absorptionTextures.clearRetainingCapacity();
		textureFogData.clearAndFree();
		textureOcclusionData.clearAndFree();
		for(textureIDs.items) |path| {
			readTextureData(path);
		}
		generateTextureArray();
	}

	pub fn generateTextureArray() void {
		const c = graphics.c;
		blockTextureArray.generate(blockTextures.items, true, true);
		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
		emissionTextureArray.generate(emissionTextures.items, true, false);
		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
		const reflectivityAndAbsorptionTextures = main.stackAllocator.alloc(Image, reflectivityTextures.items.len);
		defer main.stackAllocator.free(reflectivityAndAbsorptionTextures);
		defer for(reflectivityAndAbsorptionTextures) |texture| {
			texture.deinit(main.stackAllocator);
		};
		for(reflectivityTextures.items, absorptionTextures.items, reflectivityAndAbsorptionTextures) |reflecitivityTexture, absorptionTexture, *resultTexture| {
			const width = @max(reflecitivityTexture.width, absorptionTexture.width);
			const height = @max(reflecitivityTexture.height, absorptionTexture.height);
			resultTexture.* = Image.init(main.stackAllocator, width, height);
			for(0..width) |x| {
				for(0..height) |y| {
					const reflectivity = reflecitivityTexture.getRGB(x*reflecitivityTexture.width/width, y*reflecitivityTexture.height/height);
					const absorption = absorptionTexture.getRGB(x*absorptionTexture.width/width, y*absorptionTexture.height/height);
					resultTexture.setRGB(x, y, .{.r = absorption.r, .g = absorption.g, .b = absorption.b, .a = reflectivity.r});
				}
			}
		}
		reflectivityAndAbsorptionTextureArray.generate(reflectivityAndAbsorptionTextures, true, false);
		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));

		// Also generate additional buffers:
		if(animationSSBO) |ssbo| {
			ssbo.deinit();
		}
		if(animatedTextureSSBO) |ssbo| {
			ssbo.deinit();
		}
		if(fogSSBO) |ssbo| {
			ssbo.deinit();
		}
		animationSSBO = SSBO.initStatic(AnimationData, animation.items);
		animationSSBO.?.bind(0);

		animatedTextureSSBO = SSBO.initStaticSize(u32, animation.items.len);
		animatedTextureSSBO.?.bind(1);
		fogSSBO = SSBO.initStatic(FogData, textureFogData.items);
		fogSSBO.?.bind(7);
	}
};